### PR TITLE
[Reviewer: Richard] Fail the sprout build if plugins fail to build. 

### PR DESCRIPTION
**Pull request by AlexHockey**
Richard, please can you review this fix that makes the sprout build fail if one of the plugins fails to build. I've done this in a way that means plugin building plays nicely with `-k`.

Tested as follows:

* Clone `greeter` into the plugins directory and break it. 
* Run `make` from the top level. It exists with error code 2. 
* Run `make deb` from the top level. It exists with error code 2. 
* Clone another `greeter` into `greeter2`
* Run `make plugins-build`. This fails without building `greeter2`
* Run `make plugins-build -k`. This fails but builds `greeter2`
* Fix greeter. Run `make plugins-build`. This succeeds (exits with 0 error code)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -58,26 +58,54 @@ distclean: $(patsubst %, %_distclean, ${SUBMODULES}) sprout_distclean
...
+
+# Macro to define the rule for building a single target for a single plugin.
+#
+# Parameters:
+#   $1 - The name of the plugin to build.
+#   $2 - The make target (e.g build, test, clean, ...)
+#
+# Note that the plugins are not required to have a 'build' target, but assume
+# the default target performs a build.  For this reason we never call "make
+# build" for a plugin, and always just call "make" instead.
```
Comment (richardwhiuk) [inline Makefile]: I know this is what old code did, but is this actually sensible? Normally make on it's own defaults to `all` not `build`.

File: Makefile [hunk 2]
```diff
@@ -58,26 +58,54 @@ distclean: $(patsubst %, %_distclean, ${SUBMODULES}) sprout_distclean
...
-plugins-deb:
-	find plugins -mindepth 1 -maxdepth 1 -type d -exec ${MAKE} -C {} deb \;
-
-.PHONY: plugins-deb-only
-plugins-deb-only:
-	find plugins -mindepth 1 -maxdepth 1 -type d -exec ${MAKE} -C {} deb-only \;
-
+#
+# Plugin handling.
+#
```
Comment (richardwhiuk) [inline Makefile]: I'm undecided whether I like this style of heading. It's quite a good separator, but it also eats quite a lot of vertical space.

File: Makefile (hunk 1 above)
Comment (AlexHockey) [inline Makefile]: We discussed this and your top-level comment. The conclusion was that long-term the right thing to is come up with a build pipeline that means plugins can be built without needing a sprout checkout. Once this happens the sprout Makefile won't need knowledge of plugins at all. When/if this happens this code will be deleted so there isn't much point finessing it, especially as it's unlikely to change much.

File: Makefile (hunk 2 above)
Comment (AlexHockey) [inline Makefile]: Well with that resounding opinion I'm going to leave it as it is :-)

**Pull request merged**